### ISSUE TITLE
Add functionality to compute the Routh-Hurwitz criterion

### ISSUE THREAD

**Issue by moorepants** [sympy/sympy#26162]
Calculating symbolic (and numeric) [Routh-Hurwitz Criterion](https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion) would be beneficial for control systems analysis and could be a method of any LTI system.

Here is a very inelegant implementation:

```python
import sympy as sm


def routh_table(polynomial, var):
    p = sm.poly(polynomial, var)
    n = p.degree()
    coeffs = p.coeffs()
    table = sm.zeros(n + 1, n + 1)
    first = True
    row1, row2 = [], []
    for c in coeffs:
        if first:
            row1.append(c)
            first = False
        elif not first:
            row2.append(c)
            first = True
    for i, v in enumerate(row1):
        table[0, i] = v
    for i, v in enumerate(row2):
        table[1, i] = v
    for j in range(2, n + 1):
        for i in range(n):
            table[j, i] = (table[j - 1, 0]*table[j - 2, i + 1] -
                           table[j - 2, 0]*table[j - 1, i + 1])/table[j - 1, 0]
            table[j, i] = sm.simplify(table[j, i])
    return table
```

Example use:

```
In [3]: b0, b1, b2, b3, b4 = sm.symbols('b0, b1, b2, b3, b4')

In [4]: expr = b0*s**4 + b1*s**3 + b2*s**2 + b3*s + b4

In [5]: sm.pprint(routh_table(expr, s))
⎡            b₀               b₂  b₄  0  0⎤
⎢                                         ⎥
⎢            b₁               b₃  0   0  0⎥
⎢                                         ⎥
⎢         b₀⋅b₃                           ⎥
⎢       - ───── + b₂          b₄  0   0  0⎥
⎢           b₁                            ⎥
⎢                                         ⎥
⎢  2                                      ⎥
⎢b₁ ⋅b₄ + b₃⋅(b₀⋅b₃ - b₁⋅b₂)              ⎥
⎢───────────────────────────  0   0   0  0⎥
⎢       b₀⋅b₃ - b₁⋅b₂                     ⎥
⎢                                         ⎥
⎣            b₄               0   0   0  0⎦
```

You could return the table or just the first column of the table. Even more helpful would be to provide the inequalities in a reduced form (eliminate all multiplicative factors).

**Comment by moorepants** [comment]
Using the matrix methods is likely more elegant: https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion#Using_matrices

**Comment by virajvekaria** [comment]
> Using the matrix methods is likely more elegant: https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion#Using_matrices

By "more elegant method" do you mean this part:

https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion#:~:text=Let%20f(,is%20not%20stable.

**Comment by moorepants** [comment]
I don't know. All I know is that my code could be simplified and that there are the matrix methods that I linked.

**Comment by virajvekaria** [comment]
So I have decomposed the function and have computed the sylveser matrix, I would be integrating step 3 with 2 and then finally compute the minors of  the resultant matrix, finally giving us the solutions

**Comment by abhiphile** [comment]
I've added a PR which calculates the Routh table for any given transfer function . We can use the routh_table like this

```
from sympy import *
from sympy.physics.control.lti import routh_table
from sympy.physics.control import TransferFunction

b0, b1, b2, b3, b4, s = symbols('b0, b1, b2, b3, b4, s')
tf = TransferFunction(s**2 + 5*s + 8,1,s)
tf1 = TransferFunction(b1*s**2 + b2*s + 8,1,s)
tf2 = TransferFunction(b1*s**2 + b2*s + 8,1,s)
tf3 = TransferFunction(b1*s**3 + b2*s**2 + b3*s,s,s)
tf4 = TransferFunction(b1*s**4 + b2*s**3 + b3*s**2 + b4,1,s)
pprint(routh_table(tf))
pprint(routh_table(tf,first_col = True))
pprint(routh_table(tf1))
pprint(routh_table(tf2))
pprint(routh_table(tf3))
pprint(routh_table(tf4))
```

for these transfer functions Output will be like

```
>>> pprint(routh_table(tf))
⎡1  8  0⎤
⎢       ⎥
⎢5  0  0⎥
⎢       ⎥
⎣8  0  0⎦
>>> pprint(routh_table(tf,first_col = True))
[1, 5, 8]
>>> pprint(routh_table(tf1))
⎡b₁  8  0⎤
⎢        ⎥
⎢b₂  0  0⎥
⎢        ⎥
⎣8   0  0⎦
>>> pprint(routh_table(tf2))
⎡b₁  8  0⎤
⎢        ⎥
⎢b₂  0  0⎥
⎢        ⎥
⎣8   0  0⎦
>>> pprint(routh_table(tf3))
⎡b₁  b₃  0⎤
⎢         ⎥
⎢b₂  0   0⎥
⎢         ⎥
⎣b₃  0   0⎦
>>> pprint(routh_table(tf4))
⎡  b₁     b₃  b₄  0  0⎤
⎢                     ⎥
⎢  b₂     0   0   0  0⎥
⎢                     ⎥
⎢  b₃     b₄  0   0  0⎥
⎢                     ⎥
⎢-b₂⋅b₄               ⎥
⎢───────  0   0   0  0⎥
⎢  b₃                 ⎥
⎢                     ⎥
⎣  b₄     0   0   0  0⎦
```

**Comment by moorepants** [comment]
@abhiphile that looks good but there should be some more thought about the functinoaliy. The Routh table isn't so interesting on its own. A user would desire all unique expressions in the first column and once you have that expression you want to formulate the inequalities and possibly reduce the expressions to the simplest form we can share. Having a function that generates the table is an intermediate step, so we can likely merge that independently.

**Comment by abhiphile** [comment]
Should I add a way by which it is also possible to calculate routh_table for univariate functions.

**Comment by sylee957** [comment]
I think that we also need to discuss about handling degenerate conditions.
I see that the implementation has issues with 

```
s = Symbol('s')
routh_table(s**5 + 2*s**4 + 2*s**3 + 4*s**2 + 11*s + 10, s)
```

$$
\left[\begin{matrix}
1 & 2 & 11 & 0 & 0 & 0 \\ 
2 & 4 & 10 & 0 & 0 & 0 \\ 
0 & 6 & 0 & 0 & 0 & 0 \\ 
\tilde{\infty} & \text{NaN} & \text{NaN} & \text{NaN} & \text{NaN} & 0 \\
\text{NaN} & \text{NaN} & \text{NaN} & \text{NaN} & \text{NaN} & 0 \\
\text{NaN} & \text{NaN} & \text{NaN} & \text{NaN} & \text{NaN} & 0 
\end{matrix}\right]
$$

And also the implementation in https://github.com/sympy/sympy/pull/26170 seems like just ignoring division by zero

$$
\left[\begin{matrix}
1 & 2 & 11 & 0 & 0 & 0 \\ 
2 & 4 & 10 & 0 & 0 & 0 \\ 
0 & 6 & 0 & 0 & 0 & 0 \\ 
0 & 0 & 0 & 0 & 0 & 0 \\ 
0 & 0 & 0 & 0 & 0 & 0 \\ 
0 & 0 & 0 & 0 & 0 & 0
\end{matrix}\right]
$$

Why may just work, however, may not be mathematically sound
I think that handling degenerate cases, may be involved because we need $\epsilon$ symbol, and some infinisimal calculus.

https://control.asu.edu/Classes/MAE318/318Lecture10.pdf

(And I didn't expect this to be implemented too easy)

**Comment by moorepants** [comment]
> (And I didn't expect this to be implemented too easy)

Definitely. This needs careful thought and design. My implementation above is a naive design that can only work for very well behaved inputs.

**Comment by virajvekaria** [comment]
The matrix method listed on Wikipedia is very vague. Do you have any other resources that I can refer to?

**Comment by abhiphile** [comment]
@moorepants Can you please check the PR which solves this issue .
https://github.com/sympy/sympy/pull/26170

**Comment by moorepants** [comment]
Your PR needs to be generalized for polynomials and shouldn't have anything to do with the control module. It also needs more tests that cover all possible inputs. @sylee957 listed some special cases above, for example.

**Comment by abhiphile** [comment]
> I think that we also need to discuss about handling degenerate conditions. I see that the implementation has issues with
> 
> ```
> s = Symbol('s')
> routh_table(s**5 + 2*s**4 + 2*s**3 + 4*s**2 + 11*s + 10, s)
> ```
> 
> [12110002410000060000∞~NaNNaNNaNNaN0NaNNaNNaNNaNNaN0NaNNaNNaNNaNNaN0]
> 
> And also the implementation in #26170 seems like just ignoring division by zero
> 
> [12110002410000060000000000000000000000]
> 
> Why may just work, however, may not be mathematically sound I think that handling degenerate cases, may be involved because we need ϵ symbol, and some infinisimal calculus.
> 
> https://control.asu.edu/Classes/MAE318/318Lecture10.pdf
> 
> (And I didn't expect this to be implemented too easy)

If I take infinisimal calculus into account than I'm getting this result for the given output
``` Matrix([
[  1, 2, 11,  0, 0, 0],
[  2, 4, 10,  0, 0, 0],
[  0, 6,  0,  0, 0, 0],
[-oo, 8, -2, -2, 0, 0],
[  6, 0,  0,  0, 0, 0],
[ 10, 0,  0,  2, 0, 0]]) ```

**Comment by sylee957** [comment]
I think that you can also reference 

https://github.com/alchemyst/Dynamics-and-Control/blob/e49e69347e8e67c82d5543b5578989aca014b28f/tbcontrol/symbolic.py#L38-L65

But I warn that the license is GPL, so we shouldn't copy the code to SymPy here
However, we can take similar conceptual approach, but should do different code or have improvements
(For example, I think that simplify should be replaced with `cancel`, because that can be only relevant to cancel out some redundant factors in determinants, or that could be important to find limits with epsilon)

I am also not very sure that that code handles the degenerate case completely, so we need to verify the impementation.

I note that there are two kinds of degenerate conditions of routh table, one should also involve how to deal with complete zero rows as well.

**Comment by abhiphile** [comment]
I've written a modified version of the function which I think that it can handle both of the edge cases , Either complete row is `zero `or either first element of the first col is `zero`.

``` def routh_hurwitz_table(polynomial, var, reduce=False):
    """
    Creates the Routh-Hurwitz table for the given polynomial.

    Parameters
    ==========
    polynomial : Poly
        The Polynomial for which the Routh-Hurwitz table is to be generated.

    var : Symbol
        The variable symbol used in the polynomial.

    reduce : bool, optional
        If True, simplifies the entries in the Routh-Hurwitz table rows by cancelling
        out common terms.
        Default is False.

    Returns
    =======
    table : Matrix
        The Routh-Hurwitz table as a SymPy Matrix.

    Raises
    ======
    ValueError
        If the given polynomial is not a Poly or has a zero denominator.

    Examples
    ========
    >>> from sympy import Symbol
    >>> from sympy.polys.polytools import routh_hurwitz_table
    >>> s = Symbol('s')
    >>> exp = s**2 + 5*s + 8
    >>> routh_hurwitz_table(exp, s)
    Matrix([
    [1, 8, 0],
    [5, 0, 0],
    [8, 0, 0]])
    """
    from sympy.matrices import Matrix, matrix_multiply_elementwise, zeros
    polynomial = Poly(polynomial, var)
    if(polynomial.is_zero):
        raise ValueError("The given polynomial cannot zero.")
    n = polynomial.degree()
    coeff = polynomial.all_coeffs()
    eps = Symbol('epsilon', dummy=True)
    table = zeros(n+1,n+1)
    for j, c in enumerate(coeff):
        table[0 if j % 2 == 0 else 1, j // 2] = c

    # filling the first element in columns with eps if it is zero
    for k in range(2):
        if reduce:
            gcd_expr = gcd_list(table[k, :].tolist()[0])
            for col in range(n):
                table[k, col] = cancel(table[k, col], gcd_expr)
        if table[k, 0].is_zero:
            table[k, 0] = eps

    # filling the rest of the table
    for j in range(2,n+1):
        for i in range(n):
            table[j,i] = (table[j-1,0]*table[j-2,i+1] - table[j-2,0]*table[j-1,i+1])/table[j-1,0]
            if(i==0 and table[j,i].is_zero):
                table[j,i] = eps
        if reduce:
            gcd_expr = gcd_list(table[j, :].tolist()[0])
            for col in range(n):
                table[k, col] = cancel(table[k, col], gcd_expr)

        if table[j, :].limit(eps,0) == sympy.Matrix([[0]*(table.shape[1])]):
            diff_list = list(range(N-j, -1, -2))
            diff_list.extend([0]*(list.shape[1] - len(diff_list)))
            diff_list = Matrix([diff_list])
            table[i, :] = sympy.matrix_multiply_elementwise(diff_list, table[i-1, :])

    table = table.limit(eps,0)
    return table
```

**Comment by abhiphile** [comment]
I've collected these tests from the book .
```
    eq1 = b0*s**4 + b1*s**3 + b2*s**2 + b3*s + b4
    #Test for zero pivot element but entire row is not zero
    eq2 = s**5+3*s**4+2*s**3+6*s**2+6*s+9
    eq3 = s**6 + s**5 + 3*s**4 + 3*s**3 + 2*s**2 + 2*s + 1
    eq4 = s**4+s**3+2*s**2+2*s+1
    #Test for entire row to be zero
    eq5 = s**5+s**4+4*s**3+24*s**2+3*s+63
    #Genral tests
    eq6 = s**3 + 7*s**2 + 25*s + 39
    eq7 = s**5+2*s**4+2*s**3+4*s**2+11*s+10
```
![image](https://github.com/sympy/sympy/assets/140839576/a24e12ca-cb94-452f-9bb6-6612c4993253)
Control Systems Principles and Design by M Gopal.

**Comment by sylee957** [comment]
I think that `Poly` implicitly has symbols so you have no need to take `var` parameter, but you can use 
`polynomial.gens`

**Comment by oscarbenjamin** [comment]
For a polynomial with only rational coefficients sympy already has lots of techniques for computing or counting roots. Look through the existing methods and functionality before adding anything else e.g.:
```python
In [17]: count_roots(s**2-2, 0, oo)
Out[17]: 1
```
Another technique for the symbolic case is to use resultants to find the equation for when any eigenvalues lie on the imaginary axis:
```python
In [10]: expr = b0*s**4 + b1*s**3 + b2*s**2 + b3*s + b4
```
Assume imaginary roots of the form `s = I*y` and split equation into real and imaginary parts:
```python
In [11]: eq1, eq2 = expr.subs(s, I*y).expand().as_poly(I).all_coeffs()
```
This gives two separate equations for `y` that must be jointly satisfied. We can eliminate `y`:
```python
In [12]: resultant(eq1, eq2, y)
Out[12]: 
  2   4             2   2   2                3        4   3       3         2     2   2   2   
b₀ ⋅b₃ ⋅b₄ + 2⋅b₀⋅b₁ ⋅b₃ ⋅b₄  - 2⋅b₀⋅b₁⋅b₂⋅b₃ ⋅b₄ + b₁ ⋅b₄  - 2⋅b₁ ⋅b₂⋅b₃⋅b₄  + b₁ ⋅b₂ ⋅b₃ ⋅b₄

In [13]: resultant(eq1, eq2, y).factor()
Out[13]: 
                               2
   ⎛     2     2              ⎞ 
b₄⋅⎝b₀⋅b₃  + b₁ ⋅b₄ - b₁⋅b₂⋅b₃⎠ 
```
A root/eigenvalue can lie on the imaginary axis only if this expression is zero. As the parameters are varied continuously it is not possible for an eigenvalue to cross from the left half plane to the right half plane without this expression being zero when that eigenvalue is on the imaginary axis. In this case the condition `b4 = 0` represents an eigenvalue passing through 0 on the real line. The other condition corresponds to a complex conjugate pair crossing the imaginary axis.

The signs of the two irreducible factors then determine the placement of the eigenvalues. It is sufficient to choose example numeric values of the symbols to make each factor have a given sign. Then the number of roots in each plane can be computed numerically. Values for `b1, b2, b3` and `b4` should be chosen to represent the 4 sign combinations for the 2 factors. Then we can assign inequalities involving the factors that say when the real parts are negative.

The above assumes that the symbolic parameters are supposed to be real and that it is only the roots or eigenvalues that can be non-real. I think that is the usual setting for this sort of thing though.

**Comment by virajvekaria** [comment]
I have worked in parallel to @abhiphile on adding a function that returns the inequalities neccessary for the control system in PR #26171 as @moorepants suggested in the top-most comment, please can someone take a look at it in the PR

**Comment by moorepants** [comment]
> I have worked in parallel 

It is preferable if you work together on this. For example, you could review the first PR. Reviewers generally have little interest in dual reviewing competing PRs.

**Comment by virajvekaria** [comment]
> For a polynomial with only rational coefficients sympy already has lots of techniques for computing or counting roots. Look through the existing methods and functionality before adding anything else e.g.:
> 
> ```python
> In [17]: count_roots(s**2-2, 0, oo)
> Out[17]: 1
> ```
> 
> Another technique for the symbolic case is to use resultants to find the equation for when any eigenvalues lie on the imaginary axis:
> 
> ```python
> In [10]: expr = b0*s**4 + b1*s**3 + b2*s**2 + b3*s + b4
> ```
> 
> Assume imaginary roots of the form `s = I*y` and split equation into real and imaginary parts:
> 
> ```python
> In [11]: eq1, eq2 = expr.subs(s, I*y).expand().as_poly(I).all_coeffs()
> ```
> 
> This gives two separate equations for `y` that must be jointly satisfied. We can eliminate `y`:
> 
> ```python
> In [12]: resultant(eq1, eq2, y)
> Out[12]: 
>   2   4             2   2   2                3        4   3       3         2     2   2   2   
> b₀ ⋅b₃ ⋅b₄ + 2⋅b₀⋅b₁ ⋅b₃ ⋅b₄  - 2⋅b₀⋅b₁⋅b₂⋅b₃ ⋅b₄ + b₁ ⋅b₄  - 2⋅b₁ ⋅b₂⋅b₃⋅b₄  + b₁ ⋅b₂ ⋅b₃ ⋅b₄
> 
> In [13]: resultant(eq1, eq2, y).factor()
> Out[13]: 
>                                2
>    ⎛     2     2              ⎞ 
> b₄⋅⎝b₀⋅b₃  + b₁ ⋅b₄ - b₁⋅b₂⋅b₃⎠ 
> ```
> 
> A root/eigenvalue can lie on the imaginary axis only if this expression is zero. As the parameters are varied continuously it is not possible for an eigenvalue to cross from the left half plane to the right half plane without this expression being zero when that eigenvalue is on the imaginary axis. In this case the condition `b4 = 0` represents an eigenvalue passing through 0 on the real line. The other condition corresponds to a complex conjugate pair crossing the imaginary axis.
> 
> The signs of the two irreducible factors then determine the placement of the eigenvalues. It is sufficient to choose example numeric values of the symbols to make each factor have a given sign. Then the number of roots in each plane can be computed numerically. Values for `b1, b2, b3` and `b4` should be chosen to represent the 4 sign combinations for the 2 factors. Then we can assign inequalities involving the factors that say when the real parts are negative.
> 
> The above assumes that the symbolic parameters are supposed to be real and that it is only the roots or eigenvalues that can be non-real. I think that is the usual setting for this sort of thing though.

This is the method that Mr. Jason wanted to implement, it was I did in the PR that I mentioned above, basically to find common roots I went down the path that involves Sylvester matrices and their principal minors

**Comment by sylee957** [comment]
> For a polynomial with only rational coefficients sympy already has lots of techniques for computing or counting roots. Look through the existing methods and functionality before adding anything else e.g.:
> In [17]: count_roots(s**2-2, 0, oo)
> Out[17]: 1

I considered `count_roots` before, however, I'm not sure that that works.
The problem we are interested is counting the roots of the rectangular complex region $x + yi, x, y \in (-\infty, 0) \times (-\infty, \infty)$. So I don't think that `count_roots` with simple real line works, for example, `count_roots(x**2 + 2*x + 2, -oo, 0)`

And also, I can't get it successful for complex region as well
`count_roots(x**2 + 2*x + 2, -oo - oo*I, 0 + oo*I)` just gives `CoercionFailed: -oo is not in any domain`

I try naive substitution for infinity, and it seems like working though
`count_roots(x**2 + 2*x + 2, -9999 - 9999*I, 0 + 9999*I)`
But if we can use this, we need to find a mathematically sound

For example, it is known that there are bounds to the magnitude of the roots,
https://maa.org/sites/default/files/0746834247142.di020778.02p0335m.pdf
such that we don't really need to check the infinite region.
Maybe `count_roots` should be extended to handle this case as well

I also think that some people may be interested in other problem, for example, finding the inequality condition of (symbolic) polynomial coefficients which gives stable polynomial, rather than deciding whether the specific polynomial of numeric coefficients is stable.

For example, wikipedia article notes that $s^3 + a_2 s^2 + a_1 s + a_0$ is stable iff $a_0, a_1, a_2 > 0$ and $a_1 a_2 > a_0$

Given that symbolically parametrized control systems can be common, we also need different approach for this problem as well

**Comment by abhiphile** [comment]
Actually in my PR I've used the code at the top as a reference and implemented the two types of special conditions where the complete matrix row becomes zero and the pivot element becomes zero.

**Comment by virajvekaria** [comment]
> > I have worked in parallel
> 
> It is preferable if you work together on this. For example, you could review the first PR. Reviewers generally have little interest in dual reviewing competing PRs.

The thing is we worked on different parts of the same problem I think, because from the looks of it his code returns a routh table and my code returns the corresponding inequalities, computed by a different method which you were kind enough to link to, the matrix method from wikipedia and what @sylee957 suggested above.

**Comment by oscarbenjamin** [comment]
We can use `count_roots` if a bounded rectangle in the complex plane is given. It would be useful to extend the count roots function to handle infinite complex regions but possibly the signature is not quite right.

Various bounds are already implemented in polys and do not need to be reimplemented although methods should possibly be added to expose them through e.g. Poly. In any case a very simple bound is Lagrange's bound. For a monic polynomial it is just the maximum absolute value of the coefficients:
```python
In [30]: p = x**2 + 2*x + 2

In [31]: b = max(abs(c) for c in Poly(p, x).all_coeffs())

In [32]: count_roots(p, 0-b*I, b + b*I)
Out[32]: 0
```

> Given that symbolically parametrized control systems can be common, we also need different approach for this problem as well

I already demonstrated how to do that:
```python
In [38]: p = s**3 + a2*s**2 + a1*s + a0

In [39]: eq1, eq2 = p.subs(s, I*y).expand().as_poly(I).all_coeffs()

In [40]: resultant(eq1, eq2, y).factor()
Out[40]: 
               2
a₀⋅(a₀ - a₁⋅a₂) 
```
It suffices now to choose numeric values of `a0`, `a1` and `a2` to make each irreducible factor have a given sign and then count the roots numerically in each case. The methods in this approach have more general application than a specialised routine that tries to mimic how the Routh-Hurwitz method is used in hand calculations: you could use the same method to ask for symbolic conditions about any property of the polynomial rather than just whether the roots have negative real parts.

If existing functions like `count_roots` are not quite right for this task then it is better to improve them rather than implement something separate.

**Comment by sylee957** [comment]
> It suffices now to choose numeric values of a0, a1 and a2 to make each irreducible factor have a given sign and then count the roots numerically in each case.

I may need help following this case. I just assume that $a_0 (a_0 - a_1 a_2)^2$ is related to the answer above, however, choosing numeric values of $a_0, a_1, a_2$ seems very arbitrary for me.

**Comment by oscarbenjamin** [comment]
> choosing numeric values seems very arbitrary for me.

Suppose we want to know how many real roots this polynomial has as a function of `a`:
```python
In [42]: p = x**5 - x + a

In [43]: p
Out[43]: 
     5    
a + x  - x
```
We know that the number of real roots can only change when the discriminant is zero:
```python
In [44]: discriminant(p, x)
Out[44]: 
      4      
3125⋅a  - 256

In [45]: discriminant(p, x).factor()
Out[45]: 
      4      
3125⋅a  - 256

In [50]: r1, r2 = real_roots(discriminant(p, x).factor())

In [51]: r1.n()
Out[51]: -0.534992243981138

In [52]: r2.n()
Out[52]: 0.534992243981138
```
This partitions the set of values for `a` that we should consider into 3 cases: `a < r1`, `r1 < a < r2` and `r2 < a`. It is sufficient to pick numeric values of `a` in each case and test how many real roots there are:
```python
In [47]: len(real_roots(p.subs(a, -1)))
Out[47]: 1

In [48]: len(real_roots(p.subs(a, 0)))
Out[48]: 3

In [49]: len(real_roots(p.subs(a, 1)))
Out[49]: 1
```
We can now conclude that the number of real roots is 3 if `r1 < a < r2` and 1 otherwise. It is not necessary to check any other values of `a` because `r1` and `r2` are the only values of `a` such that the number of real roots could change. The `real_roots` function needed concrete numeric values for `a` but we have proven that is sufficient to check only 3 such values in order to make conclusions that apply for all possible values of `a`.

The same considerations happen with counting roots in each half-plane if we have an equation for when the roots can cross the imaginary axis then that partitions the possibilities and we can then pick numeric values to check each case.

Many more things can be computed with this sort of approach than just counting real roots or roots in the right half plane. This approach is in fact the basis for how cylindrical algebraic decomposition is computed which would be an extremely useful feature for sympy.

**Comment by sylee957** [comment]
I think that it may be significantly difficult problem for the cases where there are multiple parameters involved, such as finding sample points for $a_0, a_1, a_2$ as described above.
If we can restrict the problem to very simple cases like only when there is one parameter, like $K$.

**Comment by oscarbenjamin** [comment]
> I think that it may be significantly difficult problem for the cases where there are multiple parameters

This is done recursively using discriminants and resultants to eliminate one variable at a time. A function for doing this would be extremely useful.

First obtain the relevant irreducible polynomials:
```python
In [53]: a0, a1, a2 = symbols('a0, a1, a2')

In [54]: p = s**3 + a2*s**2 + a1*s + a0

In [55]: eq1, eq2 = p.subs(s, I*y).expand().as_poly(I).all_coeffs()

In [56]: factor_list(resultant(eq1, eq2, y))
Out[56]: (1, [(a₀, 1), (a₀ - a₁⋅a₂, 2)])

In [57]: facs = [fi for fi, _ in _[1]] # There should be a factors() function for this

In [59]: facs
Out[59]: [a₀, a₀ - a₁⋅a₂]
```
In the first step we will eliminate `a0` by computing discriminants and pairwise resultants:
```python
In [72]: facs2 = [discriminant(f, a0) for f in facs]

In [73]: facs2 += [resultant(facs[i], facs[j], a0) for i in range(2) for j in range(1, 2) if i != j]

In [74]: facs2
Out[74]: [1, 1, -a₁⋅a₂]

In [75]: facs2 = [f for f in facs2 if not f.is_number]

In [76]: facs2
Out[76]: [-a₁⋅a₂]
```
Now we could recursively eliminate say `a1` in the same way but we can also take a simpler approach in this case. We only care about the sign of `a1*a2` so we can take `a2` to be anything that does not make this zero, say `a2=1` and then we can have two possible values for `a1=+-1`. Taking these values back to the original factors we have:
```python
In [77]: rep1 = {a2:1, a1:-1}

In [78]: rep2 = {a2:1, a1:+1}

In [79]: [f.subs(rep1) for f in facs] + [f.subs(rep2) for f in facs]
Out[79]: [a₀, a₀ + 1, a₀, a₀ - 1]

In [81]: set().union(*(real_roots(p) for p in _))
Out[81]: {-1, 0, 1}
```
(We could handle each case separately rather than taking the union.)

Now we have four possible values of `a0` to consider and two possible values for the others. That makes 8 cases but we only need 4:
```python
In [84]: rep = [{a0:v1, a1:v2, a2:1} for v1 in [-2, -S.Half, S.Half, 2] for v2 in [-1, 1]]

In [86]: for r in rep: print(r)
{a0: -2, a1: -1, a2: 1}
{a0: -2, a1: 1, a2: 1}
{a0: -1/2, a1: -1, a2: 1}
{a0: -1/2, a1: 1, a2: 1}
{a0: 1/2, a1: -1, a2: 1}
{a0: 1/2, a1: 1, a2: 1}
{a0: 2, a1: -1, a2: 1}
{a0: 2, a1: 1, a2: 1}
```
For each of these cases we can check the sign of the polynomials:
```python
In [88]: for r in rep: print([(f, sign(f.subs(r))) for f in facs])
[(a0, -1), (a0 - a1*a2, -1)]
[(a0, -1), (a0 - a1*a2, -1)]
[(a0, -1), (a0 - a1*a2, 1)]
[(a0, -1), (a0 - a1*a2, -1)]
[(a0, 1), (a0 - a1*a2, 1)]
[(a0, 1), (a0 - a1*a2, -1)]
[(a0, 1), (a0 - a1*a2, 1)]
[(a0, 1), (a0 - a1*a2, 1)]
```
A minimal set for the four sign cases is
```python
In [90]: r2 = [rep[0], rep[2], rep[4], rep[5]]

In [91]: for r in r2: print([(f, sign(f.subs(r))) for f in facs])
[(a0, -1), (a0 - a1*a2, -1)]
[(a0, -1), (a0 - a1*a2, 1)]
[(a0, 1), (a0 - a1*a2, 1)]
[(a0, 1), (a0 - a1*a2, -1)]
```
Note that a function just for getting from `facs` to `r2` would be extremely useful. It would have far more general applicability than just the Routh-Hurwitz criterion that is discussed here. Many properties concerning inequalities with the polynomials in `facs` can be proved using `r2`.

We now have 4 candidate values `r2` for the symbols `a0`, `a1`, `a2` giving the 4 combinations for the signs of the factors in `facs`. We can compute or count the roots numerically (using `count_roots` would be more efficient):
```python
In [92]: for r2i in r2: print([r.n(3) for r in all_roots(p.subs(r2i))])
[1.21, -1.1 - 0.665*I, -1.1 + 0.665*I]
[-1.45, -0.403, 0.855]
[-1.74, 0.37 - 0.388*I, 0.37 + 0.388*I]
[-0.648, -0.176 - 0.861*I, -0.176 + 0.861*I]
```
Only the last case has all roots in the left half plane and we already computed the signs of the factors in that case so the inequalities are `a0 > 0` and `a0 - a1*a2 < 0`.

**Comment by moorepants** [comment]
>  The methods in this approach have more general application than a specialised routine that tries to mimic how the Routh-Hurwitz method is used in hand calculations: you could use the same method to ask for symbolic conditions about any property of the polynomial rather than just whether the roots have negative real parts.

Even if there is a generic method that will directly return the inequalities that guarantee  stable eigenvalues of a polynomial, I think there is still value in producing the Routh-Hurwitz style result for pedagogical purposes. Maybe then, the Routh-Hurwitz version does belong in the control package and a generic version in the polys package.

**Comment by oscarbenjamin** [comment]
> I think there is still value in producing the Routh-Hurwitz style result for pedagogical purposes.

The same result would be produced. The point is that it is better to reuse and extend or improve sympy's general functionality rather than adding many ad-hoc implementations of things like this (along with new bugs etc). Either the Routh-Hurwitz method is the best approach for this and it should be used for other things besides control or it would be better to implement this in a different way.

The first thing that could be looked at is what would be needed so that `count_roots` can check this in the numeric case and whether or not `count_roots` could be extended or improved or a similar method could be added for this case that either uses `count_roots` or perhaps is used by `count_roots`. Either `count_roots` is the right way to do this and should not be reimplemented or `count_roots` could be improved to use this method which would benefit other things as well.

SymPy has a lot of existing capability for polynomials and the existing features and API should be looked at first before implementing anything else.

**Comment by virajvekaria** [comment]
Please can someone kindly take a look at the PR that I submitted because, again, we have different approaches and do different stuff

**Comment by oscarbenjamin** [comment]
> Please can someone kindly take a look at the PR

I just tried the code from the PR and this takes an extremely long time:
```python
In [18]: p = expand((x-y)**10)

In [19]: routh_hurwitz_matrix(p, x)
...
```
By contrast this is instant:
```python
In [20]: p1, p2 = p.subs(x, I*z).as_poly(I).all_coeffs()

In [21]: p1
Out[21]: 
      9          7  3        5  5        3  7         9
- 10⋅y ⋅z + 120⋅y ⋅z  - 252⋅y ⋅z  + 120⋅y ⋅z  - 10⋅y⋅z 

In [22]: p2
Out[22]: 
 10       8  2        6  4        4  6       2  8    10
y   - 45⋅y ⋅z  + 210⋅y ⋅z  - 210⋅y ⋅z  + 45⋅y ⋅z  - z  

In [23]: factor(LC(p1, z)*LC(p2, z)*resultant(p1, p2, z))
Out[23]: 
                               101
12379400392853802748991242240⋅y 
```
This shows that the condition should be either `y < 0` or `y > 0`. A numeric check reveals which case it is:
```python
In [24]: any([re(r) >= 0 for r in all_roots(p.subs(y, -1))])
Out[24]: False

In [25]: any([re(r) >= 0 for r in all_roots(p.subs(y, 1))])
Out[25]: True
```
So `p` is stable if `y < 0`.

This is possibly not the best approach and it might be better to follow the matrix approach but with a better implementation. Actually the tabular method shown on the Wikipedia page looks vastly more efficient than computing all the minors of the Sylvester matrix.

Whichever method is used should be based on Poly, DomainMatrix etc rather than Expr and Matrix.

Anyway I think that it is a bit premature to open a PR with this yet. A good starting point would be to make something like `count_roots` that handles the purely numeric case.

**Comment by virajvekaria** [comment]
> > Please can someone kindly take a look at the PR
> 
> I just tried the code from the PR and this takes an extremely long time:
> 
> ```python
> In [18]: p = expand((x-y)**10)
> 
> In [19]: routh_hurwitz_matrix(p, x)
> ...
> ```
> 
> By contrast this is instant:
> 
> ```python
> In [20]: p1, p2 = p.subs(x, I*z).as_poly(I).all_coeffs()
> 
> In [21]: p1
> Out[21]: 
>       9          7  3        5  5        3  7         9
> - 10⋅y ⋅z + 120⋅y ⋅z  - 252⋅y ⋅z  + 120⋅y ⋅z  - 10⋅y⋅z 
> 
> In [22]: p2
> Out[22]: 
>  10       8  2        6  4        4  6       2  8    10
> y   - 45⋅y ⋅z  + 210⋅y ⋅z  - 210⋅y ⋅z  + 45⋅y ⋅z  - z  
> 
> In [23]: factor(LC(p1, z)*LC(p2, z)*resultant(p1, p2, z))
> Out[23]: 
>                                101
> 12379400392853802748991242240⋅y 
> ```
> 
> This shows that the condition should be either `y < 0` or `y > 0`. A numeric check reveals which case it is:
> 
> ```python
> In [24]: any([re(r) >= 0 for r in all_roots(p.subs(y, -1))])
> Out[24]: False
> 
> In [25]: any([re(r) >= 0 for r in all_roots(p.subs(y, 1))])
> Out[25]: True
> ```
> 
> So `p` is stable if `y < 0`.
> 
> This is possibly not the best approach and it might be better to follow the matrix approach but with a better implementation. Actually the tabular method shown on the Wikipedia page looks vastly more efficient than computing all the minors of the Sylvester matrix.
> 
> Whichever method is used should be based on Poly, DomainMatrix etc rather than Expr and Matrix.
> 
> Anyway I think that it is a bit premature to open a PR with this yet. A good starting point would be to make something like `count_roots` that handles the purely numeric case.

I have an optimization in mind, but it will still involve calculating minors in the end, do you think I should continue with this or should I do the method you suggested above?

**Comment by sylee957** [comment]
- First of all, I think that `count_roots` with infinite complex region is problematic. I think that `dup_count_roots_complex` already has . I attempted to fix this, because it is obvious that that is problem indeed, however, it was not very successful because it was difficult to sanitize the overloaded parameters `inf=None | Expr, sup=None | Expr` that used to handle the real domain. I would like to open up a new issue to do this.
  - I also notice that `count_roots` handles roots in the closed rectangle, and not the open rectangle as we expect here. `dup_count_roots_complex` has fine grained option `exclude=` to handle boundaries, however, that is not available for public yet. And again, I think that these problems can be solved from the how to disentagle bad overloaded API of `Poly.count_roots` that makes the implementation worse than how it is capable of.
- I may try to look at how to implement that projection and extension algorithm. However, I think that it would much easier if you can share the references of how your ideas came from. My observation about cylindrical algebraic decomposition are:
  - First of all, it is easy to plausibly optimize how to compute the 'projection phase', however, a lot of attempts needs verification. It looks different than a lot of references I have seen. For example, some adds 'principal subresultant coefficient' to the list of projections, and also the nonzero conditions for the coefficients of polynomials as well. Also some people tried to optimize that, but lead to errors or failure to prove them. I need to take a look at things more meticulously.
  - Second of all, I'm also alware that 'extension phase' invloves the heavy algebraic number computation. And we may not have that because algebraic numbers (like CRootOf) are really slow, or we can't even factorize the multivariate polynomials of algebraic number coefficients. This comes from the reason that the factorization should be exact, even after substitution. This is obvious from the calculus because in order to identify which region of the real univariate polynomial is positive, you need to identify the factors of even powers and odd powers. If we have to avoid algebraic numbers anyway, I would like to see how the algorithm reduces to something that uses only rational numbers, and I think that if we ignore 'boundaries', it could be plausible to find some reduced algorithm that can only use rational numbers, however, I'm also trying to find the references that really used this idea.

**Comment by oscarbenjamin** [comment]
> I think that if we ignore 'boundaries', it could be plausible to find some reduced algorithm that can only use rational numbers

If you only need to worry about strict inequalities and full-dimensional open sets then we can always choose to use rational numbers. We only need to compute RootOf to find the rational numbers in between them.

Let's discuss this somewhere else though.

**Comment by oscarbenjamin** [comment]
> I have an optimization in mind, but it will still involve calculating minors in the end, do you think I should continue with this or should I do the method you suggested above?

It is hard to say without knowing what your suggestion is. Generally the Sylvester matrix is not something that should be constructed explicitly in an actual calculation.

Have you tried the tabular method described in Wikipedia? 

The other method described there is about Sturm chains which is something that SymPy can already compute:
```python
In [2]: sturm(x**3 - 2*x**2 + x - 3)
Out[2]: 
⎡ 3      2             2            2⋅x   25         ⎤
⎢x  - 2⋅x  + x - 3, 3⋅x  - 4⋅x + 1, ─── + ──, -2079/4⎥
⎣                                    9    9          ⎦
```

**Comment by virajvekaria** [comment]
> > I have an optimization in mind, but it will still involve calculating minors in the end, do you think I should continue with this or should I do the method you suggested above?
> 
> It is hard to say without knowing what your suggestion is. Generally the Sylvester matrix is not something that should be constructed explicitly in an actual calculation.
> 
> Have you tried the tabular method described in Wikipedia?
> 
> The other method described there is about Sturm chains which is something that SymPy can already compute:
> 
> ```python
> In [2]: sturm(x**3 - 2*x**2 + x - 3)
> Out[2]: 
> ⎡ 3      2             2            2⋅x   25         ⎤
> ⎢x  - 2⋅x  + x - 3, 3⋅x  - 4⋅x + 1, ─── + ──, -2079/4⎥
> ⎣                                    9    9          ⎦
> ```

So what the method suggests on Wikipedia is doing rearrangement in rows of Sylvester matrices, I think it is possible to directly generate the rearranged matrix, and then calculate minors again.

**Comment by oscarbenjamin** [comment]
Firstly you don't need to quote an entire message when replying to it, especially if you are just replying to the message that is immediately above.

Secondly my questions were about the other methods:

Have you tried the tabular method?

What about the Sturm chain method?

I'm happy to be proved wrong but constructing the Sylvester matrix with or without rearrangement and computing minors seems almost certainly like the wrong approach to me. This is like using `det(M-l*eye(N))` to compute the characteristic polynomial of a matrix: that might be what the formula looks like in a textbook but it is definitely not the way to implement the operation computationally.
